lint/golangci-lint.bzl

```diff
@@ -0,0 +1,134 @@
+    """
+
+    # golangci-lint calls out to Go, so we need the go context
+    go = go_context(ctx)
+    inputs = srcs + [config] + go.sdk_files
+    args = ctx.actions.args()
+    args.add_all(srcs)
+
+    command = """#!/usr/bin/env bash
+        export GOROOT=$(cd "$(dirname {go_tool})/.."; pwd)
+        export GOPATH=$GOROOT
+        export GOCACHE="$(mktemp -d)"
+        export PATH="$GOPATH/bin:$PATH"
+        GOLANGCI_LINT_CACHE=$(pwd)/.cache {golangci_lint} run --config={config} $@"""
+
+    if use_exit_code:
+        command += " && touch {report}"
+    else:
+        command += " 2>{report} || true"
+
+    ctx.actions.run_shell(
+        inputs = inputs,
+        outputs = [report],
+        command = command.format(
+            golangci_lint = executable.path,
+            report = report.path,
+            config = config.path,
+            go_tool = ctx.toolchains["@io_bazel_rules_go//go:toolchain"].sdk.go.path,
+        ),
+        env = go.env,
+        arguments = [args],
+        mnemonic = _MNEMONIC,
+        tools = [executable],
+    )
+
+# buildifier: disable=function-docstring
+def _golangci_lint_aspect_impl(target, ctx):
+    if ctx.rule.kind not in ["go_binary", "go_library", "go_test"]:
+        return []
+
+    report, info = report_file(_MNEMONIC, target, ctx)
+    golangci_lint_action(ctx, ctx.executable._golangci_lint, ctx.rule.files.srcs, ctx.file._config_file, report, ctx.attr.fail_on_violation)
+    return [info]
+
+def golangci_lint_aspect(binary, config):
+    """A factory function to create a linter aspect.
+
+    Attrs:
+        binary: a golangci-lint executable.
+        config: the .golangci.yaml file
+    """
+    return aspect(
+        implementation = _golangci_lint_aspect_impl,
+        attrs = {
+            "fail_on_violation": attr.bool(),
+            "_golangci_lint": attr.label(
+                default = binary,
+                executable = True,
+                cfg = "exec",
+            ),
+            "_config_file": attr.label(
+                default = config,
+                allow_single_file = True,
+            ),
+        },
+        toolchains = ["@io_bazel_rules_go//go:toolchain"],
+    )
+
+def fetch_golangci_lint(version = "1.55.2"):
+    """Naive macro that fetches a specific version of the golangci-lint from GitHub releases, for commonly-used platforms
+
+    Args:
+        version: must be the default value. In the future this could be honored when we support multiple versions.
+    """
+    if version != "1.55.2":
+        fail("Only a single version of golangci-lint is currently mirrored. Please file an issue if you need a different version.")
+    http_archive(
+        name = "golangci_lint_linux_x86_64",
+        build_file_content = "exports_files([\"golangci-lint\"])",
+        strip_prefix = "golangci-lint-{}-linux-amd64".format(version),
+        sha256 = "ca21c961a33be3bc15e4292dc40c98c8dcc5463a7b6768a3afc123761630c09c",
```

Comment:
Isn't this problematic? If a user provides a different version as an argument to the fetch_golangci_lint function the sha256 will be wrong.

Reply:
yes, that's why it only allows the one version we currently support. We'll have to refactor this to a dictionary of hashes per release at some point.